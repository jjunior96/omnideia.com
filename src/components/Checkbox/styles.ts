import { lighten } from 'polished';
import styled, { css } from 'styled-components';

export const Container = styled.div``;

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.xsmall};
    color: ${theme.colors.gray_300};
    margin: 0;
    padding: 0;
    line-height: 1;
  `}
`;

export const Input = styled.input`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    background: transparent;
    border: 0;
    margin-right: ${theme.spacings.xxxsmall};
    outline: none;

    &:focus {
      outline: 3px solid ${lighten(0.2, theme.colors.purple_300)};
    }
  `}
`;
