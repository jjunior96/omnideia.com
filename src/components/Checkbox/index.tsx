import { InputHTMLAttributes, useState } from 'react';

import * as S from './styles';

export type CheckboxProps = {
  label?: string;
  value?: string;
  isChecked?: boolean;
  onCheck?: (status: boolean) => void;
} & InputHTMLAttributes<HTMLInputElement>;

function Checkbox({
  value,
  label,
  isChecked = false,
  onCheck,
  ...rest
}: CheckboxProps) {
  const [checked, setChecked] = useState(isChecked);

  const onChange = () => {
    const status = !checked;

    setChecked(status);

    !!onCheck && onCheck(status);
  };

  return (
    <S.Container>
      <S.Input
        id={label}
        value={value}
        type="checkbox"
        onChange={onChange}
        checked={checked}
        {...rest}
      />
      {!!label && <S.Label htmlFor={label}>{label}</S.Label>}
    </S.Container>
  );
}

export default Checkbox;
