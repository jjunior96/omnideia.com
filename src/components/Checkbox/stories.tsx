import { Story, Meta } from '@storybook/react/types-6-0';

import Checkbox, { CheckboxProps } from '.';

export default {
  title: 'Form/Checkbox',
  component: Checkbox,
  decorators: [
    (Story) => (
      <div
        style={{
          width: '40rem',
          margin: '0 auto'
        }}
      >
        <Story />
      </div>
    )
  ],
  argTypes: {
    value: {
      type: ''
    },
    onCheck: {
      action: 'checked'
    }
  }
} as Meta;

export const Default: Story<CheckboxProps> = (args) => (
  <>
    <Checkbox isChecked label="Free" {...args} />

    <Checkbox label="Half price" {...args} />

    <Checkbox label="Full" {...args} />
  </>
);
