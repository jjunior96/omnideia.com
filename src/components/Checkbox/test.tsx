import { fireEvent, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithTheme } from 'utils/tests/helpers';

import theme from 'styles/theme';

import Checkbox from '.';

describe('<Checkbox />', () => {
  it('should render the Checkbox component', () => {
    const { container } = renderWithTheme(
      <Checkbox aria-label="name-checkbox" label="Checkbox" />
    );

    const labelElement = screen.getByText(/checkbox/i);

    // Label test
    expect(labelElement).toBeInTheDocument();
    expect(labelElement).toHaveStyle({
      'font-size': '1.6rem',
      color: theme.colors.gray_300,
      margin: 0,
      padding: 0
    });

    const inputElement = screen.getByRole('checkbox', {
      name: /name-checkbox/i
    });

    // Input test
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveStyle({
      color: theme.colors.gray_300,
      background: 'transparent',
      border: '0',
      'margin-right': '0.8rem'
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should dispatch checked when user click on element', async () => {
    const onCheck = jest.fn();

    renderWithTheme(<Checkbox name="checkbox" onCheck={onCheck} />);

    // NÃ£o deve chamar a funcao onCheck
    expect(onCheck).not.toHaveBeenCalled();

    // Ao clicar no elemento, deve chamar a funcao onCheck
    fireEvent.click(screen.getByRole('checkbox'));

    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });

    expect(onCheck).toHaveBeenCalledWith(true);
  });

  it('should render with checked when isChecked passed', async () => {
    const onCheck = jest.fn();

    renderWithTheme(<Checkbox name="checkbox" onCheck={onCheck} isChecked />);

    // Ao clicar no elemento, deve chamar a funcao onCheck
    fireEvent.click(screen.getByRole('checkbox'));

    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });

    expect(onCheck).toHaveBeenCalledWith(false);
  });

  it('should be accessible with tab', () => {
    renderWithTheme(<Checkbox name="checkbox" label="Check" />);

    // No inicio, o focus deve estar no body da pagina
    expect(document.body).toHaveFocus();

    // Ao apertar em tab, o focus deve estar no checkbox
    userEvent.tab();

    expect(screen.getByRole('checkbox')).toHaveFocus();
  });
});
