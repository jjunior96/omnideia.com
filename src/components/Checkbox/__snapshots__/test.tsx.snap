// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render the Checkbox component 1`] = `
.c1 {
  font-size: 1.6rem;
  color: #3E606F;
  margin: 0;
  padding: 0;
  line-height: 1;
}

.c0 {
  color: #0D0D0D;
  background: transparent;
  border: 0;
  margin-right: 0.8rem;
  outline: none;
}

.c0:focus {
  box-shadow: 0 0 0.5rem #24A66B;
}

<div
  class="styles__Container-sc-1oopzgh-0"
>
  <input
    aria-label="name-checkbox"
    class="c0"
    id="Checkbox"
    type="checkbox"
    value=""
  />
  <label
    class="c1"
    for="Checkbox"
  >
    Checkbox
  </label>
</div>
`;
